### PR TITLE
KAFKA-14944: Reduce CompletedFetch#parseRecord() memory copy

### PR DESCRIPTION
**Pull request by LinShunKang**
This implements KIP-863: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=225152035

Direct use ByteBuffer instead of byte[] to deserialize.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java [hunk 1]
```diff
@@ -27,27 +30,45 @@
...
+    @Override
+    public String deserialize(String topic, Headers headers, ByteBuffer data) {
+        if (data == null) {
+            return null;
+        }
+
+        try {
+            if (data.hasArray()) {
+                return new String(data.array(), data.position() + data.arrayOffset(), data.remaining(), encoding);
+            } else {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: you don't really need the else here.

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java [hunk 2]
```diff
@@ -27,27 +30,45 @@
  *  value.deserializer.encoding or deserializer.encoding. The first two take precedence over the last.
  */
 public class StringDeserializer implements Deserializer<String> {
+
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: the extra line :)

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 1 above)
Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: Did you mean?
```
try {
    if (data.hasArray()) {
        return new String(data.array(), data.position() + data.arrayOffset(), data.remaining(), encoding);
    } 
    return new String(Utils.toArray(data), encoding);
} catch (UnsupportedEncodingException e) {
    throw new SerializationException("Error when deserializing ByteBuffer to string due to unsupported encoding " + encoding);
}
```

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 2 above)
Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: I see that some classes have blank line under the class, and some have no blank line, so what is the recommended style of kafka:
* have blank line: ByteArrayDeserializer, DoubleDeserializer, ListDeserializer;
* have no blank line: ByteArraySerializer, ByteBufferSerializer, LongDeserializer

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 1 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: exactly.

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 2 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: Usually I like to follow [Google's java style guide](https://google.github.io/styleguide/javaguide.html#s4.5-line-wrapping).  In this case, no new line after an opening brace.  Perhaps we can get one of the committer to comment on the styling.

**Review by kirktrue — APPROVED** [review]
LGTM.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 3]
```diff
@@ -263,6 +264,16 @@ public String deserialize(String topic, byte[] data) {
                     return super.deserialize(topic, data);
                 }
             }
+
+            @Override
+            public String deserialize(String topic, Headers headers, ByteBuffer data) {
+                if (i == recordIndex) {
+                    throw new SerializationException();
+                } else {
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:311]: you actually don't need the else there, because it will just throw.  It would be a bit cleaner if you do:
```
if (i == recordIndex) {
  throw new SerializationException();
}

i++;
return super.deserialize(topic, headers, data);
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -1413,22 +1414,19 @@ private ConsumerRecord<K, V> parseRecord(TopicPartition partition,
...
+            final K key = keyBytes == null ? null : this.keyDeserializer.deserialize(partition.topic(), headers, keyBytes);
+            final ByteBuffer valueBytes = record.value();
+            final int valueSize = valueBytes == null ? NULL_SIZE : valueBytes.remaining();
+            final V value = valueBytes == null ? null : this.valueDeserializer.deserialize(partition.topic(), headers, valueBytes);
             return new ConsumerRecord<>(partition.topic(), partition.partition(), offset,
-                                        timestamp, timestampType,
-                                        keyByteArray == null ? ConsumerRecord.NULL_SIZE : keyByteArray.length,
-                                        valueByteArray == null ? ConsumerRecord.NULL_SIZE : valueByteArray.length,
-                                        key, value, headers, leaderEpoch);
+                    timestamp, timestampType, keySize, valueSize, key, value, headers, leaderEpoch);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: consider breaking the return params into multiple lines?

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Did you mean:
```
return new ConsumerRecord<>(partition.topic(), 
    partition.partition(), 
    offset,
    timestamp, 
    timestampType, 
    keySize, 
    valueSize,
    key,
    value,
    headers,
    leaderEpoch);
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 3 above)
Comment (LinShunKang) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:311]: I just kept the same style with the previous method `public String deserialize(String topic, byte[] data)`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: right, if you think this improves the readability. :)

File: clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java [hunk 5]
```diff
@@ -60,6 +62,17 @@ default T deserialize(String topic, Headers headers, byte[] data) {
         return deserialize(topic, data);
     }
 
+    /**
+     * Deserialize a record value from a ByteBuffer into a value or object.
+     * @param topic topic associated with the data
+     * @param headers headers associated with the record; may be empty.
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java:68]: sorry, not super familiar with the Headers class, assume empty is not null? It would be good to clarify here in case if user think empty = null.

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java:68]: I copied this documentation from `deserialize(String topic, Headers headers, byte[] data)`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: hmm, it looks better this way.

**Comment by LinShunKang** [comment]
@showuon @guozhangwang @C0urante 
PTAL : )

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 2 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: We don't really have a rule for this. It's usually best practice to avoid changing a lot of spacing in a PR (if it's not required) to reduce noise.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 6]
```diff
@@ -1413,22 +1414,28 @@ private ConsumerRecord<K, V> parseRecord(TopicPartition partition,
...
-            K key = keyBytes == null ? null : this.keyDeserializer.deserialize(partition.topic(), headers, keyByteArray);
-            ByteBuffer valueBytes = record.value();
-            byte[] valueByteArray = valueBytes == null ? null : Utils.toArray(valueBytes);
-            V value = valueBytes == null ? null : this.valueDeserializer.deserialize(partition.topic(), headers, valueByteArray);
-            return new ConsumerRecord<>(partition.topic(), partition.partition(), offset,
-                                        timestamp, timestampType,
-                                        keyByteArray == null ? ConsumerRecord.NULL_SIZE : keyByteArray.length,
-                                        valueByteArray == null ? ConsumerRecord.NULL_SIZE : valueByteArray.length,
-                                        key, value, headers, leaderEpoch);
+            final long offset = record.offset();
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: We typically don't make local variables `final` in Kafka (apart from Streams), so could we undo these changes?

Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: There are similar changes in some of the other files too.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. The main changes seem to make sense to me but there's quite a few formatting changes that could be reverted to keep the PR clean and focused.

Can you also rebase on trunk to resolve the conflicts?

**Comment by mimaison** [comment]
@guozhangwang @showuon @C0urante You voted on this KIP, can you take a look? Thanks!

File: clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java (hunk 2 above)
Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/StringDeserializer.java]: > We don't really have a rule for this. It's usually best practice to avoid changing a lot of spacing in a PR (if it's not required) to reduce noise.

Ok, I got it.

**Comment by LinShunKang** [comment]
> Thanks for the PR. The main changes seem to make sense to me but there's quite a few formatting changes that could be reverted to keep the PR clean and focused.
> 
> Can you also rebase on trunk to resolve the conflicts?

Thank you for reviewing. I will address the issue.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 6 above)
Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I have resolved the conflicts and partially reverted the code formatting changes.

**Comment by showuon** [comment]
I will take a look this week. Sorry for delay

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the improvement!

@LinShunKang , I think this improvement can also apply to other types, ex: `UUIDDeserializer` is just a wrapper for `StringDeserializer`. And also, the number deserializer should also be able to be handled by ByteBuffer directly. WDYT?

**Comment by LinShunKang** [comment]
> LGTM! Thanks for the improvement!
> 
> @LinShunKang , I think this improvement can also apply to other types, ex: `UUIDDeserializer` is just a wrapper for `StringDeserializer`. And also, the number deserializer should also be able to be handled by ByteBuffer directly. WDYT?

Thank you for reviewing.
I already implement `deserialize(String, Headers, ByteBuffer)` for other Deserializers, PTAL : )

File: clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java [hunk 7]
```diff
@@ -35,4 +41,22 @@ public Double deserialize(String topic, byte[] data) {
...
+        }
+
+        if (data.remaining() != 8) {
+            throw new SerializationException("Size of data received by DoubleDeserializer is not 8");
+        }
+
+        final ByteOrder srcOrder = data.order();
+        data.order(BIG_ENDIAN);
+
+        final double value = data.getDouble(data.position());
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: @LinShunKang , why do we need to set the order to `BIG_ENDIAN` before reading it? From the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):

> Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight. 

It looks like the byte order is already considered while read from the byte buffer. Had a quick check the jdk source code, and it did check (and convert) the byte order if needed. Did I miss anything?

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: Also, the `data.position()` could be removed because by default, it'll read from current position of the byte buffer.

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: @showuon 
Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: > Also, the `data.position()` could be removed because by default, it'll read from current position of the byte buffer.

@showuon 
I use ByteBuffer.getDouble(int) instead of ByteBuffer.getDouble(), considering that ByteBuffer.getDouble(int) does not modify the offsets of the ByteBuffer.

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: @LinShunKang 
> Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].

Again, from the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):

> Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight.

The byte order is already considered while reading from the byte buffer. That means, this is not true:
> Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[]

It will read from byte[] using the current byte order (BIG_ENDIAN or LITTLE_ENDIAN). Did I miss anything?

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: > I use ByteBuffer.getDouble(int) instead of ByteBuffer.getDouble(), considering that ByteBuffer.getDouble(int) does not modify the offsets of the ByteBuffer.

I get it. We'll need the position unchanged to get the buffer size (buffer.remaining()). Thanks for the explanation.

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: > @LinShunKang
> 
> > Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].
> 
> Again, from the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):
> 
> > Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight.
> 
> The byte order is already considered while reading from the byte buffer. That means, this is not true:
> 
> > Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[]
> 
> It will read from byte[] using the current byte order (BIG_ENDIAN or LITTLE_ENDIAN). Did I miss anything?

In fact, the essence of this issue is which of the following ways we should choose for data reading:
1. Force users to write data into ByteBuffer using BIG_ENDIAN, and then we read data from ByteBuffer in BIG_ENDIAN;
2. We read data by default according to the byte order of the ByteBuffer provided by the user;

Currently, DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], so I chose method 1.

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: > Currently, DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], 

Yes, that's because we parse it by ourselves from byte array. But now, we have API provided by jdk (ByteBuffer) to allow us read according to the byte order in the buffer. So, I'm not convinced why we still need to read using `BIG_ENDIAN` order? Do you have other reason you think should stick to `BIG_ENDIAN` order?

Comment (LinShunKang) [reply clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java]: > Yes, that's because we parse it by ourselves from byte array. But now, we have API provided by jdk (ByteBuffer) to allow us read according to the byte order in the buffer. So, I'm not convinced why we still need to read using `BIG_ENDIAN` order? Do you have other reason you think should stick to `BIG_ENDIAN` order?

No, that's the only reason I chose method 1.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the update, and also improve the boolean deserializer.

**Comment by LinShunKang** [comment]
> LGTM! Thanks for the update, and also improve the boolean deserializer.

Thank you for reviewing : )

**Comment by showuon** [comment]
Failed tests are unrelated:
```
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest.putTopicStateRetriableFailure
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testSyncTopicConfigs()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest.putTopicStateRetriableFailure
    Build / JDK 17 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest.putTopicStateRetriableFailure
```

**Comment by showuon** [comment]
@LinShunKang , could you create a JIRA issue related to this KIP, and make the PR title started with `KAFKA-xxxx: Reduce...` instead of `KIP-xxx: Reduce...`. Also, the KIP needs to point to that JIRA issue in the `Status` section. You can refer to this [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API#KIP827:ExposelogdirstotalandusablespaceviaKafkaAPI-Status) . Thanks.

**Comment by LinShunKang** [comment]
> @LinShunKang , could you create a JIRA issue related to this KIP, and make the PR title started with `KAFKA-xxxx: Reduce...` instead of `KIP-xxx: Reduce...`. Also, the KIP needs to point to that JIRA issue in the `Status` section. You can refer to this [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API#KIP827:ExposelogdirstotalandusablespaceviaKafkaAPI-Status) . Thanks.

I have completed the changes mentioned above, PTAL.

**Comment by showuon** [comment]
Thanks for the improvement!

**Comment by LinShunKang** [comment]
@showuon 
Thanks for your help!
But, it seems like you've confused JIRA KAFKA-14944 with KAFKA-14945.

**Comment by showuon** [comment]
Thanks for the reminder! Updated!
